### PR TITLE
Cleanup Results helper

### PR DESCRIPTION
**Pull request by pranavkm**
* Add Results.Bytes, Results.Stream, and Results.Text overloads.
* Consolidate Physical and Virtual files helpers.
* Use default arguments to consolidate overloads.
* Remove Accepted(Uri?) / Accepted(Uri?, object? value) overload to remove ambiguity with AcceptedAt(string?, object? value);

### CODE REVIEW

**Review by Tratcher — APPROVED** [review]

File: src/Http/Http.Results/src/PublicAPI.Unshipped.txt [hunk 1]
```diff
@@ -1,105 +1,37 @@
...
-static Microsoft.AspNetCore.Http.Results.Accepted(object? value) -> Microsoft.AspNetCore.Http.IResult!
-static Microsoft.AspNetCore.Http.Results.Accepted(string? uri) -> Microsoft.AspNetCore.Http.IResult!
-static Microsoft.AspNetCore.Http.Results.Accepted(string? uri, object? value) -> Microsoft.AspNetCore.Http.IResult!
-static Microsoft.AspNetCore.Http.Results.AcceptedAtRoute(object? routeValues) -> Microsoft.AspNetCore.Http.IResult!
-static Microsoft.AspNetCore.Http.Results.AcceptedAtRoute(object? routeValues, object? value) -> Microsoft.AspNetCore.Http.IResult!
-static Microsoft.AspNetCore.Http.Results.AcceptedAtRoute(string? routeName) -> Microsoft.AspNetCore.Http.IResult!
-static Microsoft.AspNetCore.Http.Results.AcceptedAtRoute(string? routeName, object? routeValues) -> Microsoft.AspNetCore.Http.IResult!
-static Microsoft.AspNetCore.Http.Results.AcceptedAtRoute(string? routeName, object? routeValues, object? value) -> Microsoft.AspNetCore.Http.IResult!
+static Microsoft.AspNetCore.Http.Results.Accepted(string? uri = null, object? value = null) -> Microsoft.AspNetCore.Http.IResult!
```
Comment (JunTaoLuo) [inline src/Http/Http.Results/src/PublicAPI.Unshipped.txt:3]: What about Accepted(System.Uri! uri)? Are we saying just use Uri.ToString()?

File: src/Http/Http.Results/src/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -1,105 +1,37 @@
...
 static Microsoft.AspNetCore.Http.Results.Conflict() -> Microsoft.AspNetCore.Http.IResult!
 static Microsoft.AspNetCore.Http.Results.Conflict(object? error) -> Microsoft.AspNetCore.Http.IResult!
-static Microsoft.AspNetCore.Http.Results.Content(string! content) -> Microsoft.AspNetCore.Http.IResult!
-static Microsoft.AspNetCore.Http.Results.Content(string! content, Microsoft.Net.Http.Headers.MediaTypeHeaderValue? contentType) -> Microsoft.AspNetCore.Http.IResult!
-static Microsoft.AspNetCore.Http.Results.Content(string! content, string! contentType) -> Microsoft.AspNetCore.Http.IResult!
-static Microsoft.AspNetCore.Http.Results.Content(string! content, string! contentType, System.Text.Encoding! contentEncoding) -> Microsoft.AspNetCore.Http.IResult!
+static Microsoft.AspNetCore.Http.Results.Content(string! content, Microsoft.Net.Http.Headers.MediaTypeHeaderValue! contentType) -> Microsoft.AspNetCore.Http.IResult!
+static Microsoft.AspNetCore.Http.Results.Content(string! content, string? contentType = null, System.Text.Encoding? contentEncoding = null) -> Microsoft.AspNetCore.Http.IResult!
```
Comment (halter73) [inline src/Http/Http.Results/src/PublicAPI.Unshipped.txt:12]: How do we feel about just keeping Text and removing the Content overloads?

**Review by halter73 — APPROVED** [review]

File: src/Http/Http.Results/src/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (halter73) [reply src/Http/Http.Results/src/PublicAPI.Unshipped.txt:3]: I personally like less overloads for now. If people really want this, we can add the overload later.

File: src/Http/Http.Results/src/Results.cs [hunk 3]
```diff
@@ -115,63 +62,41 @@ public static IResult SignIn(ClaimsPrincipal principal, string authenticationSch
...
-        /// </summary>
-        /// <param name="content">The content to write to the response.</param>
-        /// <returns>The created <see cref="IResult"/> object for the response.</returns>
-        public static IResult Content(string content)
-            => Content(content, (MediaTypeHeaderValue?)null);
 
         /// <summary>
         /// Writes the <paramref name="content"/> string to the HTTP response.
+        /// <para>
+        /// This is an alias for <see cref="Text(string, string?, Encoding?)"/>.
```
Comment (JunTaoLuo) [inline src/Http/Http.Results/src/Results.cs:81]: I thought we were going to ignore precedence and only have Text/Stream?

**Review by JunTaoLuo — APPROVED** [review]
When tests pass.

File: src/Http/Http.Results/src/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (davidfowl) [reply src/Http/Http.Results/src/PublicAPI.Unshipped.txt:3]: They can call ToString 😄

File: src/Http/Http.Results/src/Results.cs [hunk 4]
```diff
@@ -181,11 +106,20 @@ public static IResult Content(string content, string contentType)
...
+            if (contentType is not null)
+            {
+                mediaTypeHeaderValue = MediaTypeHeaderValue.Parse(contentType);
+                mediaTypeHeaderValue.Encoding = contentEncoding ?? mediaTypeHeaderValue.Encoding;
+            }
+
+            return new ContentResult
+            {
+                Content = content,
+                ContentType = mediaTypeHeaderValue?.ToString()
```
Comment (davidfowl) [inline src/Http/Http.Results/src/Results.cs:121]: Why do we ToString after Parsing? Validation?

**Review by davidfowl — CHANGES_REQUESTED** [review]
- Should `UnprocessableEntity`/`Ok`/`NotFound` etc should haven one overload? Seems like you still have 2 method each here.
- Are Forbid/SignIn/SignOut usable with optional parameters and param arrays? 
- Do we need a status code option for Text/Bytes/Stream?
- Should `Bytes` be `Memory<byte>` instead of `byte[]`?
- Should content type be optional and assume application/octet-stream for `Bytes`/`Stream/File`?
- I think we need an overload of Stream that gives you access to the underlying response body. This is the old [PushStreamContent](https://docs.microsoft.com/en-us/previous-versions/aspnet/hh995285(v=vs.118)) from WebAPI. The scenario that comes up alot in MVC controllers is proxying a download from a blob backend. In these scenarios you want to use one of [these APIs](https://docs.microsoft.com/en-us/dotnet/api/azure.storage.blobs.specialized.blobbaseclient.downloadtoasync?view=azure-dotnet#azure-storage-blobs-specialized-blobbaseclient-downloadtoasync(system-io-stream-system-threading-can) that lets you download the blob to a Stream. The scenario is that you have a thing that's not a Stream itself but it can has a method to copy to a Stream.

```C#
app.MapGet("/blob", (BlobServiceClient client) => 
{
    BlobClient blob = client.GetBlobContainerClient("container").GetBlobClient("blob");
    return Results.Stream((stream, cancellationToken) => blob.DownloadToAsync(stream,  cancellationToken), fileDownloadName: fileName));
});
```

File: src/Http/Http.Results/src/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (Tratcher) [reply src/Http/Http.Results/src/PublicAPI.Unshipped.txt:3]: Uri.AbsoulteUri, never use ToString

File: src/Http/Http.Results/src/ObjectResult.cs [hunk 5]
```diff
@@ -55,6 +54,10 @@ public Task ExecuteAsync(HttpContext httpContext)
             }
 
             OnFormatting(httpContext);
+            if (Value is null)
```
Comment (davidfowl) [inline src/Http/Http.Results/src/ObjectResult.cs:57]: Shouldn't this go earlier?

File: src/Http/Http.Results/src/FileStreamResult.cs [hunk 6]
```diff
@@ -58,10 +36,10 @@ public FileStreamResult(Stream fileStream, MediaTypeHeaderValue contentType)
         /// </summary>
         public Stream FileStream { get; }
 
-        public Task ExecuteAsync(HttpContext httpContext)
+        public async Task ExecuteAsync(HttpContext httpContext)
         {
             var logger = httpContext.RequestServices.GetRequiredService<ILogger<FileStreamResult>>();
-            using (FileStream)
+            await using (FileStream)
```
Comment (davidfowl) [inline src/Http/Http.Results/src/FileStreamResult.cs:42]: Ah hah! That was the bug!

**Comment by halter73** [comment]
> Do we need a status code option for Text/Bytes/Stream?

I don't think so. I think 200 responses are the 99% case, and you can manually set the status on the HttpResponse if you want to

**Review by davidfowl — APPROVED** [review]

**Comment by davidfowl** [comment]
I don't get why this PR is making the blazor wasm tests fail:

```
/datadisks/disk1/work/A9DC0913/w/C4540A6A/e/Templates/BaseFolder/AspNet.qmmkzui22qe/Server/Data/ApplicationDbContext.cs(4,7): error CS0246: The type or namespace name 'IdentityServer4' could not be found (are you missing a using directive or an assembly reference?) [/datadisks/disk1/work/A9DC0913/w/C4540A6A/e/Templates/BaseFolder/AspNet.qmmkzui22qe/Server/AspNet.qmmkzui22qe.Server.csproj]
/datadisks/disk1/work/A9DC0913/w/C4540A6A/e/Templates/BaseFolder/AspNet.qmmkzui22qe/Server/Data/ApplicationDbContext.cs(13,22): error CS0246: The type or namespace name 'OperationalStoreOptions' could not be found (are you missing a using directive or an assembly reference?) [/datadisks/disk1/work/A9DC0913/w/C4540A6A/e/Templates/BaseFolder/AspNet.qmmkzui22qe/Server/AspNet.qmmkzui22qe.Server.csproj]/datadisks/disk1/work/A9DC0913/w/C4540A6A/e/Templates/BaseFolder/AspNet.qmmkzui22qe/Server/Data/ApplicationDbContext.cs(4,7): error CS0246: The type or namespace name 'IdentityServer4' could not be found (are you missing a using directive or an assembly reference?) [/datadisks/disk1/work/A9DC0913/w/C4540A6A/e/Templates/BaseFolder/AspNet.qmmkzui22qe/Server/AspNet.qmmkzui22qe.Server.csproj]
/datadisks/disk1/work/A9DC0913/w/C4540A6A/e/Templates/BaseFolder/AspNet.qmmkzui22qe/Server/Data/ApplicationDbContext.cs(13,22): error CS0246: The type or namespace name 'OperationalStoreOptions' could not be found (are you missing a using directive or an assembly reference?) [/datadisks/disk1/work/A9DC0913/w/C4540A6A/e/Templates/BaseFolder/AspNet.qmmkzui22qe/Server/AspNet.qmmkzui22qe.Server.csproj]
```

**Comment by davidfowl** [comment]
Seems like it was this https://github.com/dotnet/aspnetcore/pull/34317

**Pull request merged**